subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy

```diff
@@ -174,6 +177,213 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
+        def output = file("output.txt")
+
+        buildFile << """
+            class CustomTask extends DefaultTask {
+                @Incremental @InputFiles FileCollection brokenInputFiles
+
+                @OutputFile File output
+
+                @TaskAction execute(InputChanges changes) {
+                    output.text = changes.getFileChanges(brokenInputFiles)*.changeType
+                }
+            }
+            task inputBrokenLinkNameCollector(type: CustomTask) {
+                brokenInputFiles = files '${brokenInputFile}'
+                output = file '${output}'
+            }
+        """
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        executedAndNotSkipped ':inputBrokenLinkNameCollector'
+        output.text == "${[ADDED]}"
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        skipped ':inputBrokenLinkNameCollector'
+
+        when:
+        inputFileTarget.createFile()
+        run 'inputBrokenLinkNameCollector'
+        then:
+        executedAndNotSkipped ':inputBrokenLinkNameCollector'
+        output.text == "${[MODIFIED]}"
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        skipped ':inputBrokenLinkNameCollector'
+
+        when:
+        inputFileTarget.delete()
+        run 'inputBrokenLinkNameCollector'
+        then:
+        executedAndNotSkipped ':inputBrokenLinkNameCollector'
+        output.text == "${[MODIFIED]}"
+    }
+
+    @Issue('https://github.com/gradle/gradle/issues/9904')
+    def "task with a broken #classpathType.simpleName root input is accepted"() {
+        def brokenInputFile = file('BrokenInputFile.class').createLink("BrokenInputFile.target")
+        def output = file("output.txt")
+
+        buildFile << """
+            class CustomTask extends DefaultTask {
+                @${classpathType.name} File brokenInputFiles
+    
+                @OutputFile File output
+    
+                @TaskAction execute() {}
+            }
+            task brokenClasspathInput(type: CustomTask) {
+                brokenInputFiles = file '${brokenInputFile}'
+                output = file '${output}'
+            }
+        """
+        assert !brokenInputFile.exists()
+
+        when:
+        run 'brokenClasspathInput'
+        then:
+        executedAndNotSkipped ':brokenClasspathInput'
+
+        where:
+        classpathType << [Classpath, CompileClasspath]
+    }
+
+    @Issue('https://github.com/gradle/gradle/issues/9904')
+    def "task with a broken #classpathType.simpleName directory input is accepted"() {
+        def classes = file('classes').createDir()
+        def brokenInputFile = classes.file('BrokenInputFile.class').createLink("BrokenInputFile.target")
+        def output = file("output.txt")
+
+        buildFile << """
+            class CustomTask extends DefaultTask {
+                @${classpathType.name} File brokenInputFile
```

Comment:
It is a little bit surprising for me that the classpath property is called `brokenInputFile`, which the `brokenInputFile` variable above points to a file in the `classes` directory. Maybe this property would better be called `classes` or `classpath`?

Reply:
Renamed